internal/sink/sink.go

```diff
@@ -0,0 +1,544 @@
+func (s *Sink) deadLetter(messages ...SinkMessage) error {
+	logger := s.config.Logger.With("sink", "deadLetter")
+
+	for _, message := range messages {
+		topic := fmt.Sprintf(s.config.DeadletterTopicTemplate, message.Namespace)
+		headers := message.KafkaMessage.Headers
+		headers = append(headers, kafka.Header{Key: "error", Value: []byte(message.Error.Error())})
+
+		msg := &kafka.Message{
+			TopicPartition: kafka.TopicPartition{Topic: &topic, Partition: kafka.PartitionAny},
+			Timestamp:      message.KafkaMessage.Timestamp,
+			Headers:        headers,
+			Key:            message.KafkaMessage.Key,
+			Value:          message.KafkaMessage.Value,
+		}
+
+		err := s.producer.Produce(msg, nil)
+		if err != nil {
+			return fmt.Errorf("producing kafka message to deadletter topic: %w", err)
+		}
+	}
+
+	logger.Debug("succeeded to deadletter", "messages", len(messages))
+	return nil
+}
+
+func (s *Sink) getNamespaces() (*NamespaceStore, error) {
+	meters, err := s.config.Storage.GetMeters(s.config.Context)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get meters: %s", err)
+	}
+
+	namespaceStore := NewNamespaceStore()
+	for _, meter := range meters {
+		namespaceStore.AddMeter(meter.Namespace, meter)
+	}
+
+	return namespaceStore, nil
+}
+
+func (s *Sink) subscribeToNamespaces() error {
+	logger := s.config.Logger.With("sink", "subscribeToNamespaces")
+	ns, err := s.getNamespaces()
+	if err != nil {
+		return fmt.Errorf("failed to get namespaces: %s", err)
+	}
+
+	// Identify if we have a new namespace
+	isNewNamespace := len(ns.namespaces) != len(s.namespaceStore.namespaces)
+	if !isNewNamespace {
+		for key := range ns.namespaces {
+			if s.namespaceStore.namespaces[key] == nil {
+				isNewNamespace = true
+				break
+			}
+		}
+	}
+
+	// We always replace store to ensure we have latest meter changes
+	s.namespaceStore = ns
+
+	// We only subscribe to topics if we have a new namespace
+	if isNewNamespace {
+		// We always subscribe to all namespaces as consumer.SubscribeTopics replaces the current subscription.
+		topics := getTopics(*s.namespaceStore)
+		logger.Info("new namespaces detected, subscribing to topics", "topics", topics)
+
+		err = s.consumer.SubscribeTopics(topics, s.rebalanceCallback)
+		if err != nil {
+			return fmt.Errorf("failed to subscribe to topics: %s", err)
+		}
+		if err != nil {
+			return fmt.Errorf("failed to subscribe to topics: %s", err)
+		}
+	}
+
+	return nil
+}
+
+// Periodically flush even if MinCommitCount threshold not reached yet but MaxCommitWait time elapsed
+func (s *Sink) setFlushTimer() {
+	logger := s.config.Logger.With("sink", "flush timer")
+
+	flush := func() {
+		err := s.flush()
+		if err != nil {
+			// TODO: should we panic?
+			logger.Error("failed to flush", "err", err)
+		}
+	}
+
+	// Schedule flush
+	s.flushTimer = time.AfterFunc(s.config.MaxCommitWait, flush)
+}
+
+func (s *Sink) clearFlushTimer() {
+	if s.flushTimer != nil {
+		s.flushTimer.Stop()
+	}
+}
```

Comment:
This is technically not thread safe at the moment.

Reply:
We shouldn't have parallel flushes. Should I add a mutex?

Reply:
If there are absolutely no parallel flushes, a comment should be enough detailing that this is not concurrent safe and it's okay. Just in case we want to change things in the future.